contracts/governance/src/ZeroExVotes.sol

```diff
@@ -0,0 +1,232 @@
+        // Once we reach a single value (when low == high), we've found the right checkpoint at the index high-1, if not
+        // out of bounds (in which case we're looking too far in the past and the result is 0).
+        // Note that if the latest checkpoint available is exactly for `blockNumber`, we end up with an index that is
+        // past the end of the array, so we technically don't find a checkpoint after `blockNumber`, but it works out
+        // the same.
+        uint256 length = ckpts.length;
+
+        uint256 low = 0;
+        uint256 high = length;
+
+        if (length > 5) {
+            uint256 mid = length - Math.sqrt(length);
+            if (_unsafeAccess(ckpts, mid).fromBlock > blockNumber) {
+                high = mid;
+            } else {
+                low = mid + 1;
+            }
+        }
+
+        while (low < high) {
+            uint256 mid = Math.average(low, high);
+            if (_unsafeAccess(ckpts, mid).fromBlock > blockNumber) {
+                high = mid;
+            } else {
+                low = mid + 1;
+            }
+        }
+
+        if (high == 0) return 0;
+        else if (votingType == VotingType.Quadratic) return _unsafeAccess(ckpts, high - 1).quadraticVotes;
+        else return _unsafeAccess(ckpts, high - 1).votes;
+    }
+
+    function _writeCheckpoint(
+        Checkpoint[] storage ckpts,
+        function(uint256, uint256) view returns (uint256) op,
+        uint256 delta
+    ) private returns (uint256 oldWeight, uint256 newWeight) {
+        uint256 pos = ckpts.length;
+
+        Checkpoint memory oldCkpt = pos == 0 ? Checkpoint(0, 0) : _unsafeAccess(ckpts, pos - 1);
+
+        oldWeight = oldCkpt.votes;
+        newWeight = op(oldWeight, delta);
+
+        uint256 oldQuadraticWeight = oldCkpt.quadraticVotes;
+        uint256 newQuadraticWeight = op(oldQuadraticWeight, Math.sqrt(delta));
+
+        if (pos > 0 && oldCkpt.fromBlock == block.number) {
+            _unsafeAccess(ckpts, pos - 1).votes = SafeCast.toUint224(newWeight);
+            _unsafeAccess(ckpts, pos - 1).quadraticVotes = SafeCast.toUint224(newQuadraticWeight);
+        } else {
+            ckpts.push(
+                Checkpoint({
+                    fromBlock: SafeCast.toUint32(block.number),
+                    votes: SafeCast.toUint224(newWeight),
+                    quadraticVotes: SafeCast.toUint224(newQuadraticWeight)
+                })
+            );
+        }
+    }
+
+    function _add(uint256 a, uint256 b) public pure returns (uint256) {
+        return a + b;
+    }
+
+    function _subtract(uint256 a, uint256 b) public pure returns (uint256) {
+        return a - b;
+    }
+
+    /**
+     * @dev Access an element of the array without performing bounds check. The position is assumed to be within bounds.
+     */
+    function _unsafeAccess(Checkpoint[] storage ckpts, uint256 pos) private pure returns (Checkpoint storage result) {
+        assembly {
```

Comment:
```suggestion
        assembly ("memory-safe") {
```

without `"memory-safe"` the IR optimizer produces significantly worse code and it disables the stack limit evader

Reply:
This depends on the outcome of the discussion above https://github.com/0xProject/protocol/pull/641#discussion_r1135162016